src/GafferScene/IECoreScenePreview/PrimitiveAlgo.cpp

```diff
@@ -0,0 +1,1188 @@
+				CubicBasisf basis = sourceCurves->basis();
+				bool periodic = sourceCurves->periodic();
+				if( periodic != resultCurvesPeriodic )
+				{
+					throw IECore::Exception( "Cannot merge periodic and non-periodic curves" );
+				}
+
+				if(
+					resultCurvesBasis != invalidBasis &&
+					basis != resultCurvesBasis
+				)
+				{
+					msg( Msg::Warning, "mergePrimitives",
+						"Ignoring mismatch in curve basis and defaulting to linear"
+					);
+					resultCurvesBasis = invalidBasis;
+				}
+			}
+		}
+
+		// Process all the primvars for this primitive, adding new entries to the varInfo list, or
+		// checking that existing entries match correctly
+		for( const auto &[name, var] : primitives[i].first->variables )
+		{
+			GeometricData::Interpretation interpret = IECore::getGeometricInterpretation( var.data.get() );
+
+			IECore::TypeId varTypeId = var.data->typeId();
+
+			if( var.interpolation == PrimitiveVariable::Constant )
+			{
+				varTypeId = vectorDataTypeFromDataType( var.data.get() );
+			}
+
+			PrimVarInfo &varInfo = varInfos.try_emplace( name, varTypeId, interpret, primitives.size() ).first->second;
+
+			if( varInfo.invalid )
+			{
+				continue;
+			}
+
+			if( varTypeId == IECore::InvalidTypeId )
+			{
+				msg( Msg::Warning, "mergePrimitives",
+					fmt::format(
+						"Discarding variable \"{}\" - Cannot promote Constant primitive variable of type \"{}\".",
+						std::string( name ), var.data->typeName()
+					)
+				);
+				varInfo.invalid = true;
+				continue;
+			}
+
+			if( varInfo.typeId != varTypeId )
+			{
+				msg( Msg::Warning, "mergePrimitives",
+					fmt::format(
+						"Discarding variable \"{}\" - types don't match: \"{}\" and \"{}\"",
+						name,
+						IECore::RunTimeTyped::typeNameFromTypeId( varInfo.typeId ),
+						IECore::RunTimeTyped::typeNameFromTypeId( var.data->typeId() )
+					)
+				);
+				varInfo.invalid = true;
+				continue;
+			}
+
+			if( !varInfo.interpretInvalid )
+			{
+				if( interpret != varInfo.interpret )
+				{
+					varInfo.interpret = GeometricData::Interpretation::Numeric;
+					varInfo.interpretInvalid = true;
+					msg( Msg::Warning, "mergePrimitives",
+						fmt::format(
+							"Interpretation mismatch for primitive variable \"{}\", defaulting to \"Numeric\"", name
+						)
+					);
+				}
+			}
+
+			// For meshes and curves, the interpolation we need depends on the interpolations of all the variables
+			// being combined.
+			if( resultTypeId == IECoreScene::MeshPrimitiveTypeId )
+			{
+				switch( var.interpolation )
+				{
+					case PrimitiveVariable::Vertex:
+					case PrimitiveVariable::Varying:
+						varInfo.needsVerts = true;
+						break;
+					case PrimitiveVariable::Uniform:
+						varInfo.needsFaces = true;
+						break;
+					case PrimitiveVariable::FaceVarying:
+						varInfo.needsVerts = true;
+						varInfo.needsFaces = true;
+						break;
+					default:
+						break;
+				}
```

Comment:
Am I too optimistic in thinking that all this `needs` stuff boils down to `varInfo.interpolation = max( [ var.interpolation for var in primitives ] )`? Ah, maybe that falls apart for curves because `Varying > Vertex`?

Just trying to think of ways of limiting the type-specific stuff, or at least not having it permeate this entire function.

Reply:
It's definitely not quite that simple - between Vertex > Varying, and the combination of Varying and Uniform on a mesh requiring FaceVarying. But it is possible to implement this as a mergeInterpolations function that just takes two interpolations and the prim type - the implementation of that function is fairly messy at the moment, but it does remove a bunch of messiness from elsewhere, so it seems worth it.

Reply:
Addressed in ae07b516b9dfbba3f4cc997d2d8f2320e4dfe46a

I then tried to make the code look a bit more like a max() in this commit: 0ccada40df663d80c28e9769523021c8aced1c7a

The first commit seems like a big win ( a bit complicated looking, but moves all the complexity to one place ). The second commit might actually make things worse though - it's more concise, but I think a bit harder to understand, since it relies on completely understanding the ordering of the enum.